README.md
# Pretty modal

---

- TypeScript support
- Build with styled-components
- Light weight
- Custom styles
- CSS variables
- Accessibility
- FocusTrap
- Portal

## Installing

```shell
npm install pretty-modal
```

or

```shell
yarn add pretty-modal
```

> Note: below version 3 use default import `import Modal from 'pretty-modal'`

## Basic Example

```jsx
import React from 'react'
import {Modal} from 'pretty-modal'

const App = () => {
  const [isOpen, setIsOpen] = React.useState(false)

  return (
    <div className="App">
      <button
        onClick={() => {
          setIsOpen(true)
        }}
      >
        Open Modal
      </button>
      <Modal
        onClose={() => {
          setIsOpen(false)
        }}
        open={isOpen}
      >
        <p>Lorem, ipsum dolor sit amet consectetur adipisicing elit.</p>
        <button
          onClick={() => {
            setIsOpen(false)
          }}
        >
          Close
        </button>
      </Modal>
    </div>
  )
}

export default App
```

## CSS Variables

You can use the CSS variables for custom Style

```css
--pretty-modal-backdrop-background-color: rgba(0, 0, 0, 0.5);
--pretty-modal-backdrop-backdrop-filter: blur(1px);
--pretty-modal-backdrop-transition: all 100ms;
--pretty-modal-backdrop-transition-delay: 200ms;
--pretty-modal-backdrop-display: flex;
--pretty-modal-backdrop-align-items: center;
--pretty-modal-backdrop-justify-content: center;
--pretty-modal-backdrop-padding: 30px 0;

--pretty-modal-modal-container-transform: translateY(-10rem);
--pretty-modal-modal-container-transition: all 200ms;
--pretty-modal-modal-container-padding: 20px;
--pretty-modal-modal-container-min-height: 50px;
--pretty-modal-modal-container-min-width: 50px;
--pretty-modal-modal-container-max-width: 80%;
--pretty-modal-modal-container-max-height: 100%;
--pretty-modal-modal-container-box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
--pretty-modal-modal-container-background-color: #fff;
--pretty-modal-modal-container-border-radius: 2px;

--pretty-modal-active-transition-duration: 250ms;
--pretty-modal-active-transition-delay: 0ms;

--pretty-modal-active-modal-container-transform: translateX(0);
--pretty-modal-active-modal-container-transition-delay: 150ms;
--pretty-modal-active-modal-container-transition-duration: 350ms;
```

## Custom styles Example

```jsx
import React from 'react'
import {Modal} from 'pretty-modal'

import './App.css'

const App = () => {
  const [isOpen, setIsOpen] = React.useState(false)

  return (
    <div className="App">
      <button
        onClick={() => {
          setIsOpen(true)
        }}
      >
        Open Modal
      </button>
      <Modal
        onClose={() => {
          setIsOpen(false)
        }}
        open={isOpen}
      >
        <p>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Laudantium porro deleniti velit
        </p>
        <button
          onClick={() => {
            setIsOpen(false)
          }}
        >
          Close
        </button>
      </Modal>
    </div>
  )
}

export default App
```

`App.css`

```css
:root {
  --pretty-modal-backdrop-background-color: ; /* any color you like */
}
```

## locked Modal example

```jsx
import React, {useState} from 'react'
import {Modal} from 'pretty-modal'

const App = () => {
  const [isLocked, setIsLocked] = useState(true)
  const [isLockedOpen, setIsLockedOpen] = useState(false)

  return (
    <div>
      <button
        onClick={() => {
          setIsLockedOpen(true)
        }}
      >
        Open Locked Modal
      </button>
      <Modal
        locked={isLocked}
        onClose={() => {
          setIsLockedOpen(false)
          setIsLocked(true)
        }}
        open={isLockedOpen}
      >
        <p>s no escaping me.</p>
        <p>Once unlocked clicking outside or pressing esc will close me.</p>
        <p>
          <button
            onClick={() => {
              setIsLocked(!isLocked)
            }}
          >
            {isLocked ? 'Unlock' : 'Lock'}
          </button>
        </p>
      </Modal>
    </div>
  )
}

export default App
```

## Accessibility

```jsx
<Modal
  ariaLabelledby="modal1_label"
  ariaDescribedby="modal1_desc"
  onClose={() => {
    setIsOpen(false)
  }}
  open={isOpen}
>
  <h1 id="modal1_label">Your modal title</h1>
  <p id="modal1_desc">Your modal Description</p>
  <button
    onClick={() => {
      setIsOpen(false)
    }}
  >
    Close
  </button>
</Modal>
```

## Portal

Use `<Portal/>` separately

```jsx
import {Portal} from 'pretty-modal'
;<Portal>Hello, World</Portal>
```

## Props

| Name            | Type                | Default            | Description                                                                                                                                                                                                                                                |
| --------------- | ------------------- | ------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| children        | React.ReactNode     | `undefined`        | Where you can add the modal elements                                                                                                                                                                                                                       |
| open            | boolean             | `false`            | When it's true it will open the modal                                                                                                                                                                                                                      |
| onClose         | `() => void`        | `undefined`        | Callback fired when the Modal is requested to be closed by a click on the overlay or when user press esc key                                                                                                                                               |
| locked          | boolean             | `false`            | When it's true it will prevent close the modal when you click on the backdrop of the close button                                                                                                                                                          |
| parent          | string              | `document.body`    | You can specify the parent of the modal where you can render it.                                                                                                                                                                                           |
| parentClass     | string              | `__pretty-modal__` | You change the parent dev className                                                                                                                                                                                                                        |
| ariaLabelledby  | string              | `undefined`        | Gives the dialog an accessible name by referring to the element that provides the dialog title                                                                                                                                                             |
| ariaDescribedby | string              | `undefined`        | Gives the dialog an accessible description by referring to the dialog content that describes the primary message or purpose of the dialog.                                                                                                                 |
| `initialFocus`  | `string` or `false` | undefined          | By default, when a focus trap is activated the first element in the focus trap's tab order will receive focus. With this option you can specify a different element to receive that initial focus, or use `false` for no initially focused element at all. |


CHANGELOG.md
## [3.0.1](https://github.com/AliKdhim87/pretty-modal/compare/v3.0.0...v3.0.1) (2022-05-15)

### Bug Fixes

- **dependencies:** update dependencies ([4405c18](https://github.com/AliKdhim87/pretty-modal/commit/4405c18f1126743ee83299dd29ac56dd02aca961))

# [3.0.0](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.6...v3.0.0) (2022-02-20)

### Performance Improvements

- **version:** correcting version ([309dcab](https://github.com/AliKdhim87/pretty-modal/commit/309dcabdf68c43527829785be8a57d00d4e86907))

### BREAKING CHANGES

- **version:** previous version supposed to be breaking change

changes refer to 0613e0ed5ef206d431267c9569e97128d273b4d6

## [2.2.6](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.5...v2.2.6) (2022-02-19)

### Performance Improvements

- **modal, portal:** change to name export include portal ([0613e0e](https://github.com/AliKdhim87/pretty-modal/commit/0613e0ed5ef206d431267c9569e97128d273b4d6))

## [2.2.5](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.4...v2.2.5) (2022-02-19)

### Bug Fixes

- **package.json:** fix the types issue ([3d9ffea](https://github.com/AliKdhim87/pretty-modal/commit/3d9ffea0dfec3f1776bdea67ce374c7e32518795))

## [2.2.4](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.3...v2.2.4) (2022-02-19)

### Bug Fixes

- **package.json, gitignore:** update dependency ([18a3a7d](https://github.com/AliKdhim87/pretty-modal/commit/18a3a7df99385149476ff9b99e8d62bca68b202f))

## [2.2.3](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.2...v2.2.3) (2021-11-15)

### Bug Fixes

- **docs:** update README ([6c53280](https://github.com/AliKdhim87/pretty-modal/commit/6c5328033b3865cba8147c9d3cab2d2e65dd7fc3))

## [2.2.2](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.1...v2.2.2) (2021-11-14)

### Bug Fixes

- add unit testing ([310abdb](https://github.com/AliKdhim87/pretty-modal/commit/310abdb618a080003c5574af94eb9d2c4344efeb))

## [2.2.1](https://github.com/AliKdhim87/pretty-modal/compare/v2.2.0...v2.2.1) (2021-11-06)

### Bug Fixes

- **tsconfig:** change the target to es6 ([d08ff93](https://github.com/AliKdhim87/pretty-modal/commit/d08ff93bc7e2759b1c747427d885fa0697140050))

# [2.2.0](https://github.com/AliKdhim87/pretty-modal/compare/v2.1.2...v2.2.0) (2021-10-17)

### Features

- **initial-focus:** add initialFocus to improve the accessibility ([80eb3e6](https://github.com/AliKdhim87/pretty-modal/commit/80eb3e6a08bc80a437bdd24beda2446080ea36c0))


lib/index.cjs.js
'use strict';

var React = require('react');
var styled = require('styled-components');
var bodyScrollLock = require('body-scroll-lock');
var FocusTrap = require('focus-trap-react');
var ReactDOM = require('react-dom');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);
var FocusTrap__default = /*#__PURE__*/_interopDefaultLegacy(FocusTrap);
var ReactDOM__default = /*#__PURE__*/_interopDefaultLegacy(ReactDOM);

const Portal = ({
  children,
  parent,
  className
}) => {
  const el = React.useMemo(() => document.createElement('div'), []);
  React.useEffect(() => {
    const target = parent ? document.querySelector(parent) : document.body;
    const classList = ['__pretty-modal__'];
    if (className) className.split(' ').forEach(item => classList.push(item));
    classList.forEach(item => el.classList.add(item));
    target?.appendChild(el);
    return () => {
      target?.removeChild(el);
    };
  }, [el, parent, className]);
  return /*#__PURE__*/ReactDOM__default['default'].createPortal(children, el);
};

const GlobalStyles = styled.createGlobalStyle(["html{--pretty-modal-backdrop-background-color:rgba(0,0,0,0.5);--pretty-modal-backdrop-backdrop-filter:blur(1px);--pretty-modal-backdrop-transition:all 100ms;--pretty-modal-backdrop-transition-delay:200ms;--pretty-modal-backdrop-display:flex;--pretty-modal-backdrop-align-items:center;--pretty-modal-backdrop-justify-content:center;--pretty-modal-backdrop-padding:30px 0;--pretty-modal-modal-container-transform:translateY(-10rem);--pretty-modal-modal-container-transition:all 200ms;--pretty-modal-modal-container-padding:20px;--pretty-modal-modal-container-min-height:50px;--pretty-modal-modal-container-min-width:50px;--pretty-modal-modal-container-max-width:80%;--pretty-modal-modal-container-max-height:100%;--pretty-modal-modal-container-box-shadow:0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);--pretty-modal-modal-container-background-color:#fff;--pretty-modal-modal-container-border-radius:2px;--pretty-modal-active-transition-duration:250ms;--pretty-modal-active-transition-delay:0ms;--pretty-modal-active-modal-container-transform:translateX(0);--pretty-modal-active-modal-container-transition-delay:150ms;--pretty-modal-active-modal-container-transition-duration:350ms;}"]);
const Backdrop = styled__default['default'].div.withConfig({
  displayName: "Modal__Backdrop",
  componentId: "sc-1f8ss9b-0"
})(["position:fixed;top:0;right:0;bottom:0;left:0;background-color:var(--pretty-modal-backdrop-background-color);backdrop-filter:var(--pretty-modal-backdrop-backdrop-filter);opacity:0;transition:var(--pretty-modal-backdrop-transition);transition-delay:var(--pretty-modal-backdrop-transition-delay);display:var(--pretty-modal-backdrop-display);align-items:var(--pretty-modal-backdrop-align-items);justify-content:var(--pretty-modal-backdrop-justify-content);padding:var(--pretty-modal-backdrop-padding);& .modal-container{transform:var(--pretty-modal-modal-container-transform);transition:var(--pretty-modal-modal-container-transition);opacity:0;}&.active{transition-duration:var(--pretty-modal-active-transition-duration);transition-delay:var(--pretty-modal-active-transition-delay);opacity:1;& .modal-container{transform:var(--pretty-modal-active-modal-container-transform);opacity:1;transition-delay:var(--pretty-modal-active-modal-container-transition-delay);transition-duration:var(--pretty-modal-active-modal-container-transition-duration);}}"]);
const ModalContainer = styled__default['default'].div.withConfig({
  displayName: "Modal__ModalContainer",
  componentId: "sc-1f8ss9b-1"
})(["position:relative;padding:var(--pretty-modal-modal-container-padding);box-sizing:border-box;min-height:var(--pretty-modal-modal-container-min-height);min-width:var(--pretty-modal-modal-container-min-width);max-width:var(--pretty-modal-modal-container-max-width);max-height:var(--pretty-modal-modal-container-max-height);box-shadow:var(--pretty-modal-modal-container-box-shadow);background-color:var(--pretty-modal-modal-container-background-color);border-radius:var(--pretty-modal-modal-container-border-radius);overflow-y:auto;overflow-x:hidden;"]);

const Modal = ({
  children,
  open,
  onClose,
  locked,
  parent,
  parentClass,
  ariaDescribedby,
  ariaLabelledby
}) => {
  const [active, setActive] = React.useState(false);
  const backdrop = React.useRef(null);
  const targetRef = React__default['default'].useRef(null);
  let targetElement = null;
  React.useEffect(() => {
    const {
      current
    } = backdrop;
    targetElement = targetRef.current;

    const transitionEnd = () => setActive(open);

    const keyHandler = event => {
      if (!locked && [27].indexOf(event.which) >= 0 && onClose) {
        onClose();
        bodyScrollLock.enableBodyScroll(targetElement);
      }
    };

    const clickHandler = e => {
      if (!locked && e.target === current && onClose) {
        onClose();
        bodyScrollLock.enableBodyScroll(targetElement);
      }
    };

    if (current) {
      current && current.addEventListener('transitionend', transitionEnd);
      current.addEventListener('click', clickHandler);
      window.addEventListener('keyup', keyHandler);
    }

    if (open) {
      window.setTimeout(() => {
        document.activeElement.blur();
        setActive(open);
        bodyScrollLock.disableBodyScroll(targetElement);
      }, 10);
    }

    return () => {
      if (current) {
        current.removeEventListener('transitionend', transitionEnd);
        current.removeEventListener('click', clickHandler);
      }

      if (targetRef.current) {
        bodyScrollLock.clearAllBodyScrollLocks();
      }

      window.removeEventListener('keyup', keyHandler);
    };
  }, [open, locked, onClose]);
  return /*#__PURE__*/React__default['default'].createElement(React__default['default'].Fragment, null, /*#__PURE__*/React__default['default'].createElement(GlobalStyles, null), (open || active) && /*#__PURE__*/React__default['default'].createElement(Portal, {
    parent: parent,
    className: parentClass
  }, /*#__PURE__*/React__default['default'].createElement(FocusTrap__default['default'], null, /*#__PURE__*/React__default['default'].createElement(Backdrop, {
    ref: backdrop,
    className: active && open ? 'active' : ''
  }, /*#__PURE__*/React__default['default'].createElement(ModalContainer, {
    ref: targetRef,
    role: "dialog",
    "aria-modal": open,
    className: "modal-container",
    "aria-describedby": ariaDescribedby,
    "aria-labelledby": ariaLabelledby
  }, children)))));
};

module.exports = Modal;
//# sourceMappingURL=index.cjs.js.map
